From dc1bed039700e16029e3c3c5827cf7b262bd5999 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 19 Jan 2022 10:54:54 +0200
Subject: [PATCH] [SW-58071] gaudi3: add device specific files

This patch adds placeholder files for gaudi3 device which will
also facilitate compilation of arcboot rom for gaudi3. The updates
of address and offsets for gaudi3 will be added in subsequent
patches.

Change-Id: I40e16c83b6687dbeee6f57be55c31274a933a1d4
---

diff --git a/arch/gaudi3/arc/Makefile b/arch/gaudi3/arc/Makefile
new file mode 100644
index 0000000..573a444
--- /dev/null
+++ b/arch/gaudi3/arc/Makefile
@@ -0,0 +1,3 @@
+LDFLAGS += -static
+
+BINARIES = arcboot_rom.bin
diff --git a/arch/gaudi3/arc/arcboot_rom.lds b/arch/gaudi3/arc/arcboot_rom.lds
new file mode 100644
index 0000000..ece1742
--- /dev/null
+++ b/arch/gaudi3/arc/arcboot_rom.lds
@@ -0,0 +1,97 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
+
+OUTPUT_FORMAT("elf32-littlearc", "elf32-bigarc", "elf32-littlearc")
+OUTPUT_ARCH(arc)
+
+/* TODO: Replace with gaudi3 values */
+
+MEMORY
+{
+	ICCM	: ORIGIN = 0x00000000, LENGTH = 128K
+	STM	: ORIGIN = 0x24000000, LENGTH = 32M
+	SPI	: ORIGIN = 0x26000000, LENGTH = 16M
+	SBROM	: ORIGIN = 0x27FC0000, LENGTH = 64K
+	FWSRAM	: ORIGIN = 0x27FD0000, LENGTH = 64K
+	CS	: ORIGIN = 0x2C000000, LENGTH = 32M
+	DCCM	: ORIGIN = 0x80000000, LENGTH = 32K
+}
+
+REGION_ALIAS("REGION_IVT",	SBROM)
+REGION_ALIAS("REGION_TEXT",	SBROM)
+REGION_ALIAS("REGION_DATA",	DCCM)
+REGION_ALIAS("REGION_SDATA",	DCCM)
+REGION_ALIAS("REGION_SBSS",	DCCM)
+REGION_ALIAS("REGION_BSS",	DCCM)
+REGION_ALIAS("REGION_HEAP",	DCCM)
+REGION_ALIAS("REGION_STACK",	DCCM)
+
+ENTRY(__start)
+
+SECTIONS
+{
+	_dccm_begin = ORIGIN(DCCM);
+	_dccm_end = ORIGIN(DCCM) + LENGTH(DCCM);
+
+	.ivt : {
+		KEEP (*(.ivt));
+	} > REGION_IVT
+
+	.text : {
+		app/crt0_rom.o (.text*)
+		*(.text .text.*)
+		_etext = . ;
+	} > REGION_TEXT
+
+	.data : AT ( LOADADDR(.text) + SIZEOF(.text) ) {
+		_data = .;
+		*(.data)
+	} > REGION_DATA
+
+	.sdata : AT ( LOADADDR(.data) + SIZEOF(.data) ) {
+		__SDATA_BEGIN__ = .;
+		*(.sdata .sdata.* .gnu.linkonce.s.*)
+		_edata = .;
+		PROVIDE (edata = .);
+	} > REGION_SDATA
+
+	.sbss : {
+		__bss_start = .;
+		PROVIDE (__sbss_start = .);
+		PROVIDE (___sbss_start = .);
+		*(.dynsbss)
+		*(.sbss .sbss.* .gnu.linkonce.sb.*)
+		*(.scommon)
+		PROVIDE (__sbss_end = .);
+		PROVIDE (___sbss_end = .);
+	} > REGION_SBSS
+
+	.bss : {
+		*(.dynbss)
+		*(.bss .bss.* .gnu.linkonce.b.*)
+		*(COMMON)
+		/* Align here to ensure that the .bss section occupies space up to
+		_end. Align after .bss to ensure correct alignment even if the
+		.bss section disappears because there are no input sections. */
+		. = ALIGN(32 / 8);
+		_end = .;
+		PROVIDE (end = .);
+	} > REGION_BSS
+
+	.heap : {
+		. = ALIGN(32 / 8);
+		__start_heap = . ;
+		. = . + (DEFINED(__HEAP_SIZE) ? __HEAP_SIZE : (DEFINED(__DEFAULT_HEAP_SIZE) ? __DEFAULT_HEAP_SIZE : 5k)) ;
+		__end_heap = . ;
+		. = ALIGN(0x8);
+	} > REGION_HEAP
+
+	.stack : {
+		. = ALIGN(32 / 8);
+		__stack = . ;
+		. = . + (DEFINED(__STACK_SIZE) ? __STACK_SIZE : (DEFINED(__DEFAULT_STACK_SIZE) ? __DEFAULT_STACK_SIZE : 8k)) ;
+		__stack_top = . ;
+	} > REGION_STACK
+}
diff --git a/arch/gaudi3/arc/asic.c b/arch/gaudi3/arc/asic.c
new file mode 100644
index 0000000..f51b029
--- /dev/null
+++ b/arch/gaudi3/arc/asic.c
@@ -0,0 +1,5 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
diff --git a/arch/gaudi3/arc/asic_arcboot_rom.lds b/arch/gaudi3/arc/asic_arcboot_rom.lds
new file mode 100644
index 0000000..c03adee
--- /dev/null
+++ b/arch/gaudi3/arc/asic_arcboot_rom.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_rom.lds
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
new file mode 100644
index 0000000..666b83f
--- /dev/null
+++ b/arch/gaudi3/arc/common.c
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+#include <arcboot.h>
+
+void arcboot_rom_clear_sw_status(void) {}
+void update_warm_reset_config(void) {}
+void arcboot_rom_set_spi_write_limit(void) {}
+void pcie_phy_para_sel_set(void) {}
+void pcie_phy_set_ext_tx_rx_ctrl(void) {}
+void pcie_phy_set_protocol0_ext_tx_rx(void) {}
+void pcie_phy_set_mpll_a_b_mul_bw(__unused uint8_t separate_reference) {}
+void pcie_wait_for_sram_init_done(void) {}
+void pcie_phy_set_sram_ext_ld_done(void) {}
+void pcie_config_iatu(void) {}
+void enable_flr_handling_fsm(void) {}
+void pcie_phy_set_mpll_a_b_ssc_enable(void) {}
+
+uint32_t get_mem_cfg_mask(void)
+{
+	return 0;
+}
+
+uint8_t get_sris_mode_from_strap(void)
+{
+	return 0;
+}
+
+uint8_t get_ssc_en_from_strap(__unused uint32_t boot_strap)
+{
+	return 0;
+}
+
+uint8_t get_srns_mode_from_strap(__unused uint32_t boot_strap)
+{
+	return 0;
+}
diff --git a/arch/gaudi3/arc/dev_efuse.h b/arch/gaudi3/arc/dev_efuse.h
new file mode 100644
index 0000000..4c862c8
--- /dev/null
+++ b/arch/gaudi3/arc/dev_efuse.h
@@ -0,0 +1,38 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
+
+#ifndef _GAUDI3_DEV_EFUSE_H_
+#define _GAUDI3_DEV_EFUSE_H_
+
+/* TODO: Replace with gaudi3 values */
+
+#define EFUSE_KEY_HASH_OFFSET				0x14
+
+/*
+ * 00: ROM, 01: EROM, 10: RSVD, 11:ROM
+ */
+#define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION		13
+#define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_MASK	0x3
+#define EFUSE_SYSCFG_RECOVERY_CODE_LOCATION_EROM	0x1
+
+#define EFUSE_SYSCFG_ENABLE_TPM_MASK			BIT(11)
+#define EFUSE_SYSCFG_FW_VERSION_LOCATION		7
+#define EFUSE_SYSCFG_FW_VERSION_LOCATION_MASK		0x3
+#define EFUSE_SYSCFG_KEY_LOCATION			5
+#define EFUSE_SYSCFG_KEY_LOCATION_MASK			0x3
+#define EFUSE_SYSCFG_REVOKATION_LOCATION		7
+#define EFUSE_SYSCFG_REVOKATION_LOCATION_MASK		0x3
+
+#define EFUSE_SYSCFG_FW_VERSION_LOCATION_SPI		BIT(8)
+#define EFUSE_SYSCFG_REVOKATION_LOCATION_SPI		BIT(8)
+
+/*
+ * As per recommendation from Security Architect, if FUSE is not available,
+ * card should not continue in the secure boot mode. Hence the default
+ * can be 0 (with no enabling of the Signature of Validation also.
+ */
+#define EFUSE_SYSCFG_DEFAULT				0
+
+#endif /* _GAUDI3_DEV_EFUSE_H_ */
diff --git a/arch/gaudi3/arc/dev_security.h b/arch/gaudi3/arc/dev_security.h
new file mode 100644
index 0000000..719c06f
--- /dev/null
+++ b/arch/gaudi3/arc/dev_security.h
@@ -0,0 +1,42 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
+
+#ifndef _GAUDI3_DEV_SECURITY_H_
+#define _GAUDI3_DEV_SECURITY_H_
+
+#include <sha512.h>
+
+/* TODO: Replace with gaudi3 values */
+
+#define NUM_KEYS_IN_ROM					5
+#define NUM_KEYS_IN_EFUSE				3
+#define SECURITY_DWORD_PER_KEY				12
+#define SECURITY_KEY_HASH_COUNT				3
+#define NUM_DWORDS_AVLBL_IN_PSOC_SEC			8
+
+enum security_key_location {
+	KEY_LOCATION_ROM,
+	KEY_LOCATION_EFUSE
+};
+
+/* size of key properties in bytes */
+#define KEY_PROP_RR_SZ					384
+#define KEY_PROP_MOD_SZ					384
+#define KEY_PROP_EXP_SZ					8
+#define KEY_PROP_INV_SZ					4
+#define KEY_PROP_NB_SZ					4
+#define KEY_PROP_HDR_SZ					8
+
+/* 384 + 384 + 8 + 4 + 4 = 784 bytes */
+#define PUB_KEY_SZ		(KEY_PROP_RR_SZ + KEY_PROP_MOD_SZ + \
+				KEY_PROP_EXP_SZ + KEY_PROP_INV_SZ + \
+				KEY_PROP_NB_SZ)
+
+/* 784 + 8 = 792 */
+#define TOTAL_PUB_KEY_PROPS_SZ  (PUB_KEY_SZ + KEY_PROP_HDR_SZ)
+
+extern const uint8_t rom_key_hashes[NUM_KEYS_IN_ROM][SHA384_SUM_LEN];
+
+#endif /* _GAUDI3_DEV_SECURITY_H_ */
diff --git a/arch/gaudi3/arc/onix.c b/arch/gaudi3/arc/onix.c
new file mode 100644
index 0000000..f51b029
--- /dev/null
+++ b/arch/gaudi3/arc/onix.c
@@ -0,0 +1,5 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
diff --git a/arch/gaudi3/arc/onix_arcboot_rom.lds b/arch/gaudi3/arc/onix_arcboot_rom.lds
new file mode 100644
index 0000000..8564bf3
--- /dev/null
+++ b/arch/gaudi3/arc/onix_arcboot_rom.lds
@@ -0,0 +1,12 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
+
+INCLUDE arcboot_rom.lds
+
+/* This is SRAM address, used for Onix targets */
+MEMORY
+{
+	SBROM	: ORIGIN = 0x11000000, LENGTH = 64K
+}
diff --git a/arch/gaudi3/arc/pldm.c b/arch/gaudi3/arc/pldm.c
new file mode 100644
index 0000000..f51b029
--- /dev/null
+++ b/arch/gaudi3/arc/pldm.c
@@ -0,0 +1,5 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
diff --git a/arch/gaudi3/arc/pldm_arcboot_rom.lds b/arch/gaudi3/arc/pldm_arcboot_rom.lds
new file mode 100644
index 0000000..c03adee
--- /dev/null
+++ b/arch/gaudi3/arc/pldm_arcboot_rom.lds
@@ -0,0 +1,6 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
+
+ INCLUDE arcboot_rom.lds
diff --git a/arch/gaudi3/arc/secureboot.c b/arch/gaudi3/arc/secureboot.c
new file mode 100644
index 0000000..b030246
--- /dev/null
+++ b/arch/gaudi3/arc/secureboot.c
@@ -0,0 +1,12 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+#include <arcboot.h>
+
+int secureboot_verify_image(__unused uint32_t image_size, __unused uint32_t image_offset)
+{
+	return 0;
+}
diff --git a/arch/gaudi3/arc/security.c b/arch/gaudi3/arc/security.c
new file mode 100644
index 0000000..f75fdd8
--- /dev/null
+++ b/arch/gaudi3/arc/security.c
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier:     GPL-2.0+
+/*
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ * All Rights Reserved.
+ */
+
+#include <arcboot.h>
+
+/* TODO: Add gaudi3 specific implementation */
+
+void security_i2c_s_rr_hbw_config(__unused uint8_t spi_controller_en) {}
+void security_smi_rr_hbw_config(__unused uint8_t spi_controller_en) {}
+void security_jtag_rr_hbw_config(__unused uint8_t spi_controller_en) {}
+void security_disable_jtag_if(void) {}
+void security_handle_twi(void) {}
+void security_load_keys(void) {}
+void arcboot_rom_request_cpld_wdog(void) {}
+void security_enable_jtag_if(void) {}
+void security_jtag_dap(__unused uint8_t status) {}
+void check_and_enable_4b_termination(void) {}
+void security_debug_if_shared(void) {}
+void security_pcie_rr_config(__unused uint8_t spi_en) {}
diff --git a/arch/gaudi3/arc/soc.h b/arch/gaudi3/arc/soc.h
new file mode 100644
index 0000000..378e9f6
--- /dev/null
+++ b/arch/gaudi3/arc/soc.h
@@ -0,0 +1,56 @@
+/* SPDX-License-Identifier:     GPL-2.0+
+ *
+ * Copyright (C) 2022 HabanaLabs Ltd.
+ */
+
+#ifndef _SOC_H_
+#define _SOC_H_
+#include <gaudi3_regs.h>
+
+/* TODO: Replace with gaudi3 values */
+
+#define CPU_FREQ			500000000
+#define CPU_FREQ_US			500	/* CPU_FREQ / 1000000 */
+
+#define FW_SIZE				65535
+#define SIG_SIZE			384
+#define KEY_SIZE			784
+#define BLOB_START			4
+
+#define SPI_FLASH_BLOCK_SIZE		1024
+
+/* 2MB for Gaudi2 */
+#define FLASH_VALID_MASK		0x1FFFFF
+#define FLASH_VALID_OFFSET		0
+#define FLASH_VALID_MAX_SIZE		0x200000
+
+#define HL_SRAM_BASE			0x1000007FF0000000ULL
+#define HL_SP_SRAM_BASE			0x1000007FF7FE0000ULL
+#define HL_SRAM_PHY_BASE		0x1000FFFFFD000000ULL
+
+/*
+ * This is not SDRAM Base, but the base address of General region
+ * into which the rest of the components are mapped
+ */
+#define HL_SDRAM_BASE			0x20000000
+#define HL_SRAM_LOG_BASE		0x11000000
+
+#define HL_FLASH_OFFSET			0x6000000
+#define HL_ICCM_OFFSET			0x7040000
+#define HL_SBROM_OFFSET			0x7FC0000
+#define HL_FWSRAM_OFFSET		0x7FD0000
+#define HL_SCRATCHPAD_SRAM_OFFSET	0x7FE0000
+#define HL_PCIE_FWSRAM_OFFSET		0x7FF0000
+#define HL_CFG_BASE			0xF8000000
+#define HL_ICCM_BASE			(HL_SRAM_BASE + HL_ICCM_OFFSET)
+
+#define HL_ICCM_SIZE			0x20000
+#define HL_SP_SRAM_SIZE			0x10000
+#define HL_SRAM_PHY_SIZE		0x2FFFFF
+
+#define HL_ICCM_END			(HL_ICCM_BASE + HL_ICCM_SIZE - 1)
+#define HL_SP_SRAM_END			(HL_SP_SRAM_BASE + HL_SP_SRAM_SIZE - 1)
+#define HL_SRAM_PHY_END			(HL_SRAM_PHY_BASE +\
+					HL_SRAM_PHY_SIZE - 1)
+
+#endif /* _SOC_H_ */
